restapi/auth.go

```diff
@@ -11,38 +13,85 @@ import (
 	"github.com/uwblueprint/shoe-project/config"
 	"github.com/uwblueprint/shoe-project/internal/database/models"
 	"github.com/uwblueprint/shoe-project/restapi/rest"
-	"golang.org/x/crypto/bcrypt"
-	"gorm.io/gorm"
 )
 
+func generateStateOauthCookie(w http.ResponseWriter) (string, error) {
+	duration, err := config.GetTokenExpiryDuration()
+	if err != nil {
+		return "", err
+	}
+	var expiration = time.Now().Add(duration)
+	secure := (config.GetMode() == config.MODE_PROD)
+
+	b := make([]byte, 16)
+	rand.Read(b)
+	state := base64.URLEncoding.EncodeToString(b)
+	cookie := http.Cookie{Name: "oauthstate", Value: state, Expires: expiration, HttpOnly: true, Secure: secure}
+	http.SetCookie(w, &cookie)
+
+	return state, nil
+}
+
 func (api api) Login(w http.ResponseWriter, r *http.Request) render.Renderer {
-	var user models.User
-	err := json.NewDecoder(r.Body).Decode(&user)
+	state, err := generateStateOauthCookie(w)
+	if err != nil {
+		return rest.ErrInternal(api.logger, err)
+	}
+
+	url := api.oauthconfig.AuthCodeURL(state)
+	http.Redirect(w, r, url, http.StatusTemporaryRedirect)
+
+	return rest.JSONStatusOK(url)
+}
+
+func (api api) AuthCallback(w http.ResponseWriter, r *http.Request) render.Renderer {
+	expectedState, _ := r.Cookie("oauthstate")
+	if expectedState.Value != r.FormValue("state") {
```

Comment:
Just so I understand the `r.FormValue("state")` is added on by google right? We don't have to include anything on our frontend?

Reply:
Same for `r.FormValue("code")` below right?

Reply:
yep, this callback endpoint is only ever called by google; all the query params are set by them too